### PR TITLE
Add `locale` option

### LINKED ISSUE #72: Case conversion with specific locale
Case conversion are done with `String.prototype.toLocaleLowerCase()` and `String.prototype.toLocaleUpperCase()` without explicit locale parameter and this causes host dependent output.
Example:
```js
camelCase('lorem-ipsum')
//=> 'loremIpsum' if host locale is en-US
//=> 'loremİpsum' if host locale is tr-TR
```
This causes unexpected results int hosts with non en-US environment.

I'd suggest to extend `options` parameter to have a `locale` property to be passed to aforementioned methods.

```js
camelCase('lorem-ipsum', {locale: 'en-US'});
//=> 'loremIpsum' if host locale is en-US
//=> 'loremIpsum' if host locale is tr-TR
```

### PR DESCRIPTION
**Pull request by rasgele**
Fixes #72 
Added locale property to option parameter along with tests and doc.

### CODE REVIEW

**Comment by Qix-** [comment]
Very much :+1:. This is the proper solution here. @sindresorhus I know your stance on adding options, but I think this is a valid exception to that rule.

Could you potentially add tests for cases where the output differs between locales?

**Comment by rasgele** [comment]
@Qix- there are such tests in `test.js`. Also samples in readme.md and `index.d.ts`.

File: test.js [hunk 1]
```diff
@@ -64,6 +64,10 @@ test('camelCase', t => {
 	t.is(camelCase('桑德在这里。'), '桑德在这里。');
 	t.is(camelCase('桑德在这里。'), '桑德在这里。');
 	t.is(camelCase('桑德_在这里。'), '桑德在这里。');
+	t.is(camelCase('lorem-ipsum', {locale: 'tr-TR'}), 'loremİpsum');
+	t.is(camelCase('lorem-ipsum', {locale: 'en-EN'}), 'loremIpsum');
+	t.is(camelCase('lorem-ipsum', {locale: ['tr', 'TR', 'tr-TR']}), 'loremİpsum');
+	t.is(camelCase('lorem-ipsum', {locale: ['en-EN', 'en-GB']}), 'loremIpsum');
```
Comment (sindresorhus) [inline test.js:70]: These should be in a separate test.

File: readme.md [hunk 2]
```diff
@@ -71,6 +77,13 @@ Default: `false`
 
 Uppercase the first character: `foo-bar` → `FooBar`
 
+##### locale
+
+Type: `string | string[]`\
+Default: `undefined`
+
+Locale to be used with lowercase or uppercase conversions.
```
Comment (sindresorhus) [inline readme.md:85]: This needs to explicitly say what the expected string locale format it. (for example, what `en-US` refers to).

Comment (sindresorhus) [reply readme.md:85]: It's also unclear why one would want to specify multiple locales and how that works.

File: readme.md [hunk 3]
```diff
@@ -71,6 +77,13 @@ Default: `false`
 
 Uppercase the first character: `foo-bar` → `FooBar`
 
+##### locale
+
+Type: `string | string[]`\
+Default: `undefined`
```
Comment (sindresorhus) [inline readme.md:83]: I assume the default is the user's locale? If so, that should be the default here.

File: index.d.ts [hunk 4]
```diff
@@ -51,6 +57,19 @@ camelCase(['foo', 'bar']);
...
 //=> 'FooBar'
+
+camelCase('lorem-ipsum', {locale: 'en-US'});
+//=> 'loremIpsum'
+
+camelCase('lorem-ipsum', {locale: 'tr-TR'});
+//=> 'loremİpsum'
+
+camelCase('lorem-ipsum', {locale: ['en-US', 'en-GB]});
+//=> 'loremIpsum'
```
Comment (sindresorhus) [inline index.d.ts:69]: The readme and index.d.ts should be fully in sync.

File: index.d.ts [hunk 5]
```diff
@@ -51,6 +58,18 @@ camelCase(['foo', 'bar']);
 
 camelCase(['__foo__', '--bar'], {pascalCase: true});
 //=> 'FooBar'
+
+camelCase('lorem-ipsum', {locale: 'en-US'});
+//=> 'loremIpsum'
```
Comment (sindresorhus) [inline index.d.ts:63]: It's enough with this example here. The this and the other examples can be in the doc comment for the option.

File: test.js (hunk 1 above)
Comment (rasgele) [reply test.js]: done.

File: readme.md (hunk 3 above)
Comment (rasgele) [reply readme.md]: Yes, updated as suggested.

File: readme.md (hunk 2 above)
Comment (rasgele) [reply readme.md]: I updated with an explicit reference to [String.prototype.toLocaleUpperCase()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toLocaleUpperCase) and excerpt from that page.  It explains format and usage with multiple locale.
Details (like expected format being [BCP 47](https://tools.ietf.org/html/bcp47) etc.) seemed too much in this context, but it  is up to you.

File: index.d.ts (hunk 4 above)
Comment (rasgele) [reply index.d.ts]: I am a little confused about this. I interpreted your next comment as "Keep a single example in `index.d.ts` and move others to `readme.md` file", which breaks sync.

File: index.d.ts (hunk 5 above)
Comment (rasgele) [reply index.d.ts:63]: Kept single example here. All examples are in doc comment for `locale` option.

Comment (sindresorhus) [reply index.d.ts:63]: > All examples are in doc comment for locale option.

I think you forgot to add it there. Also add those to the readme (`locale` option part).

Comment (rasgele) [reply index.d.ts:63]: Sorry, I forgot. 
There are 4 examples in total. All goes under 'Usage' section or 'API' section? I think 'API' section is appropriate as all examples are similar format (`camelcase(....)`).

File: readme.md (hunk 2 above)
Comment (sindresorhus) [reply readme.md]: One example in Usage and all in the API docs for `locale`.

Comment (rasgele) [reply readme.md]: done.

**Comment by sindresorhus** [comment]
> The readme and index.d.ts should be fully in sync.

The `locale` option description differs between the readme and index.d.ts

**Comment by rasgele** [comment]
> > The readme and index.d.ts should be fully in sync.
> 
> The `locale` option description differs between the readme and index.d.ts

Fixed.

File: readme.md [hunk 6]
```diff
@@ -71,6 +74,27 @@ Default: `false`
...
+
+camelCase('lorem-ipsum', {locale: 'tr-TR'});
+//=> 'loremİpsum'
+
+camelCase('lorem-ipsum', {locale: ['en-US', 'en-GB']});
+//=> 'loremIpsum'
+
+camelCase('lorem-ipsum', {locale: ['tr', 'TR', 'tr-TR']});
+//=> 'loremİpsum'
+```
```
Comment (sindresorhus) [inline readme.md:96]: > The readme and index.d.ts should be fully in sync.

File: index.d.ts [hunk 7]
```diff
@@ -6,6 +6,13 @@ declare namespace camelcase {
 		@default false
 		*/
 		readonly pascalCase?: boolean;
+
+		/**
+		From `String.prototype.toLocaleUpperCase()`: The locale parameter indicates the locale to be used to convert to upper/lower case according to any locale-specific case mappings. If multiple locales are given in an Array, the best available locale is used. The default locale is the host environment’s current locale.
+
+		@default The host environment’s current locale
```
Comment (sindresorhus) [inline index.d.ts:13]: This is incorrect syntax. `@default` only works with a JS value.

File: index.d.ts [hunk 8]
```diff
@@ -6,6 +6,13 @@ declare namespace camelcase {
 		@default false
 		*/
 		readonly pascalCase?: boolean;
+
+		/**
+		From `String.prototype.toLocaleUpperCase()`: The locale parameter indicates the locale to be used to convert to upper/lower case according to any locale-specific case mappings. If multiple locales are given in an Array, the best available locale is used. The default locale is the host environment’s current locale.
```
Comment (sindresorhus) [inline index.d.ts:11]: > The default locale is the host environment’s current locale.

Is specified twice.

Comment (sindresorhus) [reply index.d.ts:11]: I don't really see the point in mentioning `From `String.prototype.toLocaleUpperCase()`:`.

Comment (rasgele) [reply index.d.ts]: Removed the reference and double default specification.

**Comment by gorkemyontem** [comment]
Hi @sindresorhus, @rasgele any progress on this PR? I'm also having problems with Jest. 
Thanks!

**Comment by sindresorhus** [comment]
There's still feedback that needs to be addressed.

File: readme.md (hunk 6 above)
Comment (rasgele) [reply readme.md:96]: readme contains 4 examples and index.d.ts contains one example due to your previous comment: https://github.com/sindresorhus/camelcase/pull/75#discussion_r456964680
Do you means smth else?

File: index.d.ts (hunk 7 above)
Comment (rasgele) [reply index.d.ts:13]: Then I append following to the description of `locale` parameter:
"If not provided, host environment's current locale is used". 

Do you have other suggestion for such a case?

**Comment by rasgele** [comment]
Sorry, I missed latest comments. 
Now I replied and based on @sindresorhus's responses I will try to push changes quickly.

File: index.d.ts (hunk 7 above)
Comment (sindresorhus) [reply index.d.ts:13]: You can just write `Default: The host environment’s current locale.`

File: readme.md (hunk 6 above)
Comment (sindresorhus) [reply readme.md:96]: These 4 examples should be here: https://github.com/sindresorhus/camelcase/pull/75/files#diff-b52768974e6bc0faccb7d4b75b162c99R14 (Just like the readme).

File: index.d.ts (hunk 7 above)
Comment (rasgele) [reply index.d.ts]: done.

File: readme.md (hunk 6 above)
Comment (rasgele) [reply readme.md:96]: Oh, ok. thanks. Done.

File: index.d.ts (hunk 5 above)
Comment (rasgele) [reply index.d.ts:78]: So, I guess this is also resolved.

**Comment by rasgele** [comment]
@sindresorhus Employed 'request a review' as a 'ping'... sounded appropriate :)

**Pull request merged**